### PR TITLE
[#7532][feat] AutoDeploy: Move the logit selection before the final Gemm

### LINKED ISSUE #7532: [Feature]:  Review logits post-processing performance
### 🚀 The feature, motivation and pitch

@lucaslie found differences between AD's and PT's logits post-processing implementations and we should study these to understand the perf implications of the 2 implementations.

Slack: https://nvidia.slack.com/archives/C08T55LHSG4/p1756943076165769

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by nvchenghaoz**
This PR fixes the #7532, the goal is to move the logit selection before the final GEMM to reduce the overall computation for the final GEMM for the prefill / mixed stage. 

Also integrate the code to handle the case where the transform is not enabled. 

## Release Notes

* **New Features**
  * Added gather_logits_before_lm_head transform that gathers hidden states for selected logits before the language model head computation, enabling new inference execution paths.

* **Bug Fixes**
  * Fixed logits tensor handling in the inference engine to properly squeeze batch dimensions when logits are already in tensor format.

* **Tests**
  * Added comprehensive unit tests for the gather_logits_before_lm_head transform supporting both generate and packed input formats.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by nvchenghaoz** [comment]
@nvpohanh , we discussed a bit in the meeting and Yoco mentioned he might work on that. I added the transform here because this is a necessary perf optimization for nemotron MOE model.

**Review by lucaslie — COMMENTED** [review]
a little too much vibe code ;)

File: tensorrt_llm/_torch/auto_deploy/transform/library/gather_logits_before_lm_head.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+        hidden_states: Hidden states tensor [batch, seq_len, hidden] or
+            [1, total_tokens, hidden] or [total_tokens, hidden]
+        logit_gather_ids: Gather indices [max_batch_size], one per sequence
+            (index of token to gather)
+
+    Returns:
+        Gathered hidden states [batch, hidden] for generate, [1, max_batch_size, hidden] for packed
+    """
+    # Generate format: [batch, 1, hidden] -> seq_len == 1
+    # Packed format: [1, total_tokens, hidden] -> seq_len > 1
```
[the remark this answers is not in the document]
Comment (nvchenghaoz) [reply tensorrt_llm/_torch/auto_deploy/transform/library/gather_logits_before_lm_head.py]: yeah, it could be mixed prefill + decode or pure prefill.

**Comment by lucaslie** [comment]
started a rebased/combined branch with my https://github.com/NVIDIA/TensorRT-LLM/pull/9764 changes but it's still work in progress: https://github.com/nv-auto-deploy/TensorRT-LLM/tree/ll/meta_logits_combined

The rebase actually was too big to resolve so I am kind of starting from scratch
